### PR TITLE
Cleanup extract_meshes

### PR DESCRIPTION
**Pull request by re0312**
# Objective

- clean up extract_mesh_(gpu/cpu)_building

## Solution

- gpu_building no need to hold  `prev_render_mesh_instances`
- using `insert_unique_unchecked` instead of simple insert as we know all entities are unique
- direcly get `previous_input_index ` in par_loop 

## Performance
this should also bring a slight performance win.

cargo run --release --example many_cubes --features bevy/trace_tracy -- --no-frustum-culling     
`extract_meshes_for_gpu_building`

image

### CODE REVIEW

File: crates/bevy_pbr/src/render/mesh.rs [hunk 1]
```diff
@@ -710,96 +706,49 @@ pub fn extract_meshes_for_gpu_building(
...
-                previous_input_index: match previous_input_index {
-                    Some(previous_input_index) => previous_input_index.into(),
-                    None => u32::MAX,
-                },
-            }) as u32;
-
-            // Record the [`RenderMeshInstance`].
-            render_mesh_instances.insert(
-                entity,
+        for ((e, s), u) in queue.drain(..) {
```
Comment (pcwalton) [inline crates/bevy_pbr/src/render/mesh.rs:743]:
```suggestion
        for ((entity, shared), mesh_uniform) in queue.drain(..) {
```

File: crates/bevy_pbr/src/render/mesh.rs [hunk 2]
```diff
@@ -710,96 +706,49 @@ pub fn extract_meshes_for_gpu_building(
...
+            let buffer_index = current_input_buffer.push(u);
+            let translation = Vec3::new(u.transform[0].w, u.transform[1].w, u.transform[2].w);
```
Comment (pcwalton) [inline crates/bevy_pbr/src/render/mesh.rs:745]:
```suggestion
            let translation = mesh_uniform.transform.row(3).xyz();
```

File: crates/bevy_pbr/src/render/mesh.rs [hunk 3]
```diff
@@ -710,96 +706,49 @@ pub fn extract_meshes_for_gpu_building(
...
+            render_mesh_instances.insert_unique_unchecked(
+                e,
                 RenderMeshInstanceGpu {
-                    translation: builder.transform.translation,
-                    shared: builder.shared,
-                    current_uniform_index: NonMaxU32::try_from(current_uniform_index)
-                        .unwrap_or_default(),
+                    shared: s,
```
Comment (pcwalton) [inline crates/bevy_pbr/src/render/mesh.rs:749]:
```suggestion
                    shared,
```

File: crates/bevy_pbr/src/render/mesh.rs [hunk 4]
```diff
@@ -710,96 +706,49 @@ pub fn extract_meshes_for_gpu_building(
...
-            }) as u32;
-
-            // Record the [`RenderMeshInstance`].
-            render_mesh_instances.insert(
-                entity,
+        for ((e, s), u) in queue.drain(..) {
+            let buffer_index = current_input_buffer.push(u);
+            let translation = Vec3::new(u.transform[0].w, u.transform[1].w, u.transform[2].w);
+            render_mesh_instances.insert_unique_unchecked(
+                e,
```
Comment (pcwalton) [inline crates/bevy_pbr/src/render/mesh.rs:747]:
```suggestion
                entity,
```

**Review by pcwalton — APPROVED** [review]
Looks good with a few nits, thanks!

**Review by james7132 — APPROVED** [review]

**Review by mockersf — APPROVED** [review]

**Pull request merged**